src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -44,12 +44,19 @@ Status CoreWorkerDirectTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
           task_spec,
           [this, actor_id, task_id](Status status, const rpc::CreateActorReply &reply) {
             if (status.ok()) {
-              RAY_LOG(DEBUG) << "Created actor, actor id = " << actor_id;
-              // Copy the actor's reply to the GCS for ref counting purposes.
-              rpc::PushTaskReply push_task_reply;
-              push_task_reply.mutable_borrowed_refs()->CopyFrom(reply.borrowed_refs());
-              task_finisher_->CompletePendingTask(
-                  task_id, push_task_reply, reply.actor_address());
+              if (reply.creation_cancelled()) {
```

Comment:
Why don't we just handle this in `if (!status.ok())` code path? Why do we need a special `creation_cancelled `? (I think if the actor scheduling is already cancelled by because the actor is already killed, it should go that path?)

Reply:
I initially went that route, ended up going with a cancellation field for a few reasons:
1. We need to be able to delineate between actor creation failure (which should be retried) and actor creation cancellation (which should not be retried), and there wasn't an existing status code that seemed like a good delineator, so we'd need to add a new one.
2. The operation didn't fail, it was determined to not be needed and was therefore canceled, which feels like a valid part of the success path in the actor creation protocol, so returning a non-ok status felt weird. E.g., we don't return a non-ok status when a worker lease request is cancelled: https://github.com/ray-project/ray/blob/0b39285dea6b4266c05a52652678db5f2939f813/src/ray/protobuf/node_manager.proto#L81

Could you say more about why you think this should fall in the `!status.ok()` path?

Reply:
status.ok() sounded like it means the actor creation is succeeded. But it seems like the definition of status is not clearly defined within a callback, so it might be okay. 

Still prefer to have a status like SchedulingCancelled error and handle it within !status.ok() path, but if this pattern is already used at other places, it might be okay (still don't prefer it...)

Reply:
>status.ok() sounded like it means the actor creation is succeeded. But it seems like the definition of status is not clearly defined within a callback, so it might be okay.

Our mixture of transport-level and application-level statuses, piggybacking off of the gRPC status rather than keeping them separate and using the rich error model or the protobuf payload for the latter, has always bothered me. So I think I have a subconscious aversion to using it for anything other than transport-level failures. 😂 

>Still prefer to have a status like SchedulingCancelled error and handle it within !status.ok() path, but if this pattern is already used at other places, I am okay with the current style

I'll take a stab at this and see how it feels!